---
title: "Data_analysis_nifH"
output:
  html_notebook:
    toc: yes
    toc_depth: 4
    toc_float: true
  html_document: null
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, tidy = TRUE, highlight = TRUE)
```

```{r, echo = F, message=FALSE}
library(knitr)
library(tidyverse)
library(car)
library(broom)
library(gridExtra)
library(phyloseq)
library(DESeq2)
#library(GGally)
#library(parallel)
```

# experimental design:

![experimental set-up](Experimental_layout.jpg)

In the experimental design, each diversity treatments consist of a meta-community ("big core") of four separate sediment communities ("small core") that correspond to the habitats.

For habitat level 1, each **metacommunity** is replicated 4 times for each habitat (resulting in 16 meta-communities in total)

habitat level 2 to 4, consist of four replicated meta-communities each, and each meta-community contains 2x2 [level 2], (2x1 + 1x2) [level 3] and 4x1 [level 4] habitat communities ("small cores"), respectively. 

There are two missing samples

+ Level 3 - replicate A - habitat cyano
+ Level 3 - replicate D - habitat silt


# import and clean metadata and functional data

```{r, message = FALSE}
meta <- read_tsv("metadata.txt")

N_func <- read_tsv("summer_habdiv_n2fix_deni_new.txt")
qPCR <- read_tsv("qPCR.txt")
Par_frct <- read.table("Fraction_Parallogues_nifh.txt", sep ="\t", header = T)
Chl_frct <- read.table("frct_Chl.txt")



FuncDat <- meta %>% 
  filter(season == "summer") %>% 
  select(-season) %>% 
  mutate(habitat = ifelse(Level == 1, sediment, "mix")) %>% 
  select(-habitat) %>% 
  left_join(N_func) %>% 
  left_join(qPCR) %>% 
  left_join(Par_frct) %>% 
  left_join(Chl_frct) %>% 
  arrange(Level, replicate, sediment)

kable(head(FuncDat))
```

both, the 16S and the nifH primer picked up more then the target genes.

+ The nifh primers picked up a sizable fraction (`r paste(paste(round(min(FuncDat$frct.par, na.rm = T)*100),  round(max(FuncDat$frct.par, na.rm = T)*100), sep = " - "), "%")`) of nifh - parallogues, i.e. close relatives to nifh sequences that are not involved in nitrogen fixation. 

+ The 16S primers picked up chloroplast sequences (`r paste(paste(round(min(FuncDat$frct_Chl, na.rm = T)),  round(max(FuncDat$frct_Chl, na.rm = T)), sep = " - "), "%")`)

As we used the same primers in the qPCR assays, we need to account for that. We therefore correct each sample for the % of non-target sequences is in different runs. This correction is of cause not perfect as we don't know the variability of the % of target sequences picked up in different runs, as the qPCR conditions varied slightly from the pcr conditions for library preparation and of cause as the sequence count from amplicon sequencing is only roughly quantitative

```{r}

FuncDat <- FuncDat %>% 
  mutate(nifH = nifH * (1 - frct.par)) %>% 
  mutate(`16S` = `16S` * ((100 - frct_Chl) / 100))

```



# Hypothesis 1

**Habitat diversity leads to increased bacterial diversity and to increased functional guild diversity (i.e. both more bacterial species in general but also a greater diversity of species carrying the genes for nitrogen fixation and denitrification respectively)**

To test this hypothesis I break it down into two question:

## Hypothesis 1 a)

**does functional gene diveristy increase with overall bacterial diversity?**

We can test this hypothesis using the alpha diversity measurements in the *small* cores, with all 50 data points. However, the measurements taken from small cores that came from the same *big* core are not independent and we need to account for that. Also, we measured *three* functional genes. Because the absolute values of the functional gene diversities are not (necessarily) comparable, we test each response variable separately. 

The (3) model tests the following hypothesis

+ does functional gene diversity change habitat (is the mean diversity different between the habitats)
+ is functional gene diversity correlated with bacterial diversity (overall effect)
+ is the relationship between bacterial diversity and functional gene diversity dependent on the habitat (significant interaction effect)


### import & reshape diveristy data

note that I calculated three diversity metrics, *richness*, *effective number of species* and *effective number of phylogenetic equidistinct species*. Richness is a hard to estimate and therefore unreliable diversity metric and the ecological relevance of richness can be debated. Therefore I report it but exclude it from the data analysis. Effective number of species and phylogenetic diversity tend to be well correlated (see Diversity estimation), wherefore it is unlikely to yield very different results. Effective number of *species* (otus) is easier to interpret (number of species in an equally diverse but completely even community) the the phylogenetic diversity (number of unique "phylotypes", i.e. number of species in an equally phylodiverse community where all species have the same abundance and are equally unrelated to each other). To minimize the overall number of linear models, we present phylogentic diversity but only use the effective number of species in the data analysis. 



```{r}
DIV <- read.table("Diversity_estimations.txt")

DIVeffN <- DIV %>%
  select(-contains("S_")) %>% 
  select(-contains("y_")) %>% 
  select(-contains("phylo")) %>% 
  gather(func_gen, func_div, matches("_nifh|_nosz1|_nosz2")) %>% 
  na.omit() %>% 
  left_join(meta)
  

kable(head(DIVeffN))
```

### test main effect of big cores

note that the data structure could imply non-independence of the small cores that are grouped within each big core. To account for this we first test if there is a main effect of *big cores* 

```{r}
DIVeffN %>% 
  filter(Level > 1) %>%
  lm(func_div ~ func_gen * MC.ID, data=.) %>% 
  anova()
```

This should give us the [type I sum of squares](http://goanna.cs.rmit.edu.au/~fscholer/anova.php) and hence whether the big cores (MC.ID) have a significant effect *after* accounting for functional gene. I conclude that as neither the main effect nor the interaction is remotely significant we don't need to worry about non-independence and can safely pool the data. 


As the absolute diversity values of three functional genes are not necessarily comparable, we model them separately (which also allows us to transform the data separately as appropriate)

### nifH
```{r, fig.width = 12, fig.height = 10}
mod_nifh <- lm(func_div ~  sediment * effN_bac, data = filter(DIVeffN, func_gen == "effN_nifh")) 

par(mfrow=c(3,3))
plot(mod_nifh)
hist(resid(mod_nifh))

shapiro.test(residuals(mod_nifh))
ncvTest(mod_nifh)
```
Here and for all subsequent models we use three methods to evaluate whether we meet the assumption for linear models (i.e. normality and heteroscedasticity (unequal variances) of residuals)

+ we inspect the diagnostic plots
+ we test for non-normality using the Shapiro-Wilk test of normality (see `?shapiro.test`)
+ we test for unequal variance using the Score Test for Non-Constant Error Variance (Breusch-Pagan test). (see `?ncvTest`)

Here, the diagnostic plots look fine and both tests are non-significant. We proceed to interpret the model

Also not that as our design is unbalanced (we have different number of observation for the different factor levels) we need to think about which type sum of squares to use. 

I adopt the following strategy: 

+ first we test for a significant interaction (default SS, type I)
+ if no significant interaction is found, we test the significance of the main effects with type II sum of squares. This tests the first main effect *over* the second main effect and vice versa, i.e. the effect for (A|B) and (B|A). It equivalent to the test performed for the *second* main effect in type I sum of squares. 
+ if the interaction is significant (conventionally defined as p â‰¤ 0.05) we test the main effects with type III sum of squares, i.e. each main effect after accounting for the other main effect **and** the interaction, i.e. (A | B,A*B) & (B | A,A\*B). However, if the interaction is significant the main effects are hard to interpret. 

#### results
```{r}
# type I sum of squares testing the interaction
anova(mod_nifh)

#type II sum of squares testing the main effects
Anova(mod_nifh, type = 2)

mod_nifh %>% tidy %>% kable

mod_nifh %>% glance %>% kable

PRED <- predict(mod_nifh, type = "terms")
Const <- attr(PRED,"constant")

# predict
PRED.df <-  data.frame(y = PRED) %>% 
  mutate(y.effN_bac = y.effN_bac + Const) %>% 
  mutate(y.sediment = y.sediment + Const) %>% 
  mutate(y.sediment.effN_bac = y.sediment.effN_bac + y.effN_bac + y.sediment - Const) %>% 
  mutate(x = as.vector(DIVeffN[DIVeffN$func_gen == "effN_nifh", ]$effN_bac)) %>% 
  mutate(sediment = as.vector(DIVeffN[DIVeffN$func_gen == "effN_nifh", ]$sediment))

plot_nifh <- DIVeffN %>% 
  filter(func_gen == "effN_nifh") %>% 
  ggplot(aes(x = effN_bac, y = func_div, colour = sediment) )+
  geom_point()+
  geom_line(data = PRED.df, aes(x = x, y = y.sediment.effN_bac), linetype = "dashed")+
  geom_line(data = PRED.df, aes(x = x, y = y.effN_bac), colour = "black")+
  theme_bw()+
  labs(x = "effective number of bacterial otus", y = "effective number of nifH otus", title="nifH")+
  theme(legend.position = "bottom")+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))

plot_nifh
```

+ because the interaction is non-significant we went on and tested the main effects (type II sum of squares)

+ the main effect of 'sediment' is highly significant. The mean diversity changes with sediment with silt having the highest mean diversity and sand the lowest. 

+ the main effect of bacterial diversity is significant, too.  


### nosZ I


```{r, fig.width = 12, fig.height = 10}
mod_nosz1 <- lm(func_div ~ sediment * effN_bac, data = filter(DIVeffN, func_gen == "effN_nosz1")) 

par(mfrow=c(3,3))
plot(mod_nosz1)
hist(resid(mod_nosz1))

shapiro.test(residuals(mod_nosz1))
ncvTest(mod_nosz1)

```

The diagnostic plots don't look great and the tests suggest both non-normality and heteroscedasticity

I try to log-transform the nosZ1 data 


```{r, fig.width = 12, fig.height = 10}
mod_nosz1_log <- lm(log(func_div, base = 10) ~  sediment * effN_bac, data = filter(DIVeffN, func_gen == "effN_nosz1")) 

par(mfrow=c(3,3))
plot(mod_nosz1_log)
hist(resid(mod_nosz1_log))

shapiro.test(residuals(mod_nosz1_log))
ncvTest(mod_nosz1_log)
```


The plots look okay now and the tests are unsignificant. The transformation was successful 

#### results
```{r}
# type I type of squares. Only for testing the interaction
anova(mod_nosz1_log)

#type III sum of squares testing the main effects (as we have a significant interaction)
Anova(mod_nosz1_log, type = 3)

mod_nosz1_log %>% tidy %>%  mutate(estimate = 10^estimate) %>% kable

mod_nosz1_log %>% glance %>% kable

PRED <- predict(mod_nosz1_log, type = "terms")
Const <- attr(PRED,"constant")

# predict
PRED.df <-  data.frame(y = PRED) %>% 
  mutate(y.sediment.effN_bac = 10^(y.sediment.effN_bac + y.effN_bac + y.sediment + Const)) %>% 
  mutate(y.effN_bac = 10 ^ (y.effN_bac + Const)) %>% 
  mutate(y.sediment = 10 ^ (y.sediment + Const)) %>% 
  mutate(x = as.vector(DIVeffN[DIVeffN$func_gen == "effN_nosz1", ]$effN_bac)) %>% 
  mutate(sediment = as.vector(DIVeffN[DIVeffN$func_gen == "effN_nosz1", ]$sediment))

plot_noszI <- DIVeffN %>% 
  filter(func_gen == "effN_nosz1") %>% 
  ggplot(aes(x = effN_bac, y = func_div, colour = sediment) )+
  geom_point()+
  geom_line(data = PRED.df, aes(x = x, y = y.sediment.effN_bac), linetype = "dashed")+
  geom_line(data = PRED.df, aes(x = x, y = y.effN_bac), colour = "black")+
  theme_bw()+
  labs(x = "effective number of bacterial otus", y = "effective number of nosZ I otus", title="nosZ I")+
  theme(legend.position = "bottom")+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))

plot_noszI

```

+ there is a significant (although week) interaction. We test the main effects with type III sum of squares
+ mean diversity of nosZ I otus changes with sediment, although the effect depends on the bacterial diversity. Highest diversity is observed in silt and lowest diversity in sand and ruppia
+ there is no effect of the effective number of total bacterial otus on the effective number of nosZ I otus


### nosZ II

```{r, fig.width = 12, fig.height = 10}
mod_nosz2 <- lm(func_div ~ sediment * effN_bac, data = filter(DIVeffN, func_gen == "effN_nosz2")) 

par(mfrow=c(3,3))
plot(mod_nosz2)
hist(resid(mod_nosz2))

shapiro.test(residuals(mod_nosz2))
ncvTest(mod_nosz2)


```

The diagnostic plots and the tests look fine

#### results
```{r}

anova(mod_nosz2)
Anova(mod_nosz2, type = 3)

mod_nosz2 %>% tidy %>%  kable

mod_nosz2 %>% glance %>% kable

PRED <- predict(mod_nosz2, type = "terms")
Const <- attr(PRED,"constant")

# predict
PRED.df <-  data.frame(y = PRED) %>% 
  mutate(y.effN_bac = y.effN_bac + Const) %>% 
  mutate(y.sediment = y.sediment + Const) %>% 
  mutate(y.sediment.effN_bac = y.sediment.effN_bac + y.effN_bac + y.sediment - Const) %>% 
  mutate(x = as.vector(DIVeffN[DIVeffN$func_gen == "effN_nosz2", ]$effN_bac)) %>% 
  mutate(sediment = as.vector(DIVeffN[DIVeffN$func_gen == "effN_nosz2", ]$sediment))

plot_noszII <- DIVeffN %>% 
  filter(func_gen == "effN_nosz2") %>% 
  ggplot(aes(x = effN_bac, y = func_div, colour = sediment) )+
  geom_point()+
  geom_line(data = PRED.df, aes(x = x, y = y.sediment.effN_bac), linetype = "dashed")+
  geom_line(data = PRED.df, aes(x = x, y = y.effN_bac), colour = "black")+
  theme_bw()+
  labs(x = "effective number of bacterial otus", y = "effective number of nosZ II otus", title="nosZ II")+
  theme(legend.position = "bottom")+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))

plot_noszII
```

+ there is a week but significant interaction. The relationship between bacterial diversity and nosZ II otu diversity depends on the sediment. It is strongest in sediments cyano and sand and flat in ruppia silt. The main effects must therefore be interpreted with care.

+ there is a strong significant positive relationship between the effective number of nosZ II otus and the overall bacterial diversity after accounting for the main effect of sediment and the interaction

+ there is a week significant difference between the sediments



## Hypothesis 1 b)

**Does habitat diversity leads to increased bacterial gamma diversity and to increased functional guild gamma diversity**

*gamma diversity* has been calculated on the level of the large cores, comprising 4 different small cores with 1 - 4 different habitats. Note that for level 1 alpha and gamma diversity are the same (only 1 pooled sample was taken for each identical habitat). For level 3, the habitats were weighted by there abundance (i.e. the habitat that was present twice was weighted double)

+ We take habitat as categorical variable
+ We exclude two level 3 cores with missing samples
+ We test each functional gene separately
+ We use effective number of species as diversity metric

The first possibility how habitat diversity could impact functional gene diversity is through it's impact on bacterial 16S diversity. We can hence test if 16S gamma diversity is at all related to habitat diversity. 

### reshape diveristy data


```{r}

DIVeffN_y <- DIV %>%
  select( Level, MC.ID, contains("y_")) %>% 
  select( - matches("_S_|_phylo")) %>% 
  distinct(.keep_all = TRUE) %>% 
  gather(func_gen, y_func_div, matches("_nifh|_nosz1|_nosz2")) %>% 
  na.omit() 
  

kable(head(DIVeffN_y))
```


### 16S gamma diversity

```{r, fig.width = 12, fig.height = 10}

mod_16S <- lm(y_effN_bac ~ Level, data = filter(DIVeffN_y, func_gen == "y_effN_nosz1")) 

par(mfrow=c(3,3))
plot(mod_16S)
hist(resid(mod_16S))

shapiro.test(residuals(mod_16S))
ncvTest(mod_16S)

```

The residuals are approximately normal but have unequal variances. However it's not too bad and transforming the data doesn't help so we go with it. 

#### results
```{r}
anova(mod_16S)

PRED <- 
  data.frame(y = predict(mod_16S)) %>% 
  mutate(Level = as.vector(DIVeffN_y[DIVeffN_y$func_gen == "y_effN_nosz2", ]$Level))

plot_16S_y <- DIVeffN_y %>% 
  filter(func_gen == "y_effN_nosz2") %>% 
  ggplot(aes(x = Level, y = y_effN_bac, colour = as.factor(Level)))+
  geom_point()+
  geom_line(data = PRED, aes(x = Level, y = y, group = 1), colour = "black", size = 0.3)+
  theme_bw()+
  labs(x = "bacterial gamma diversity", y = "nosZ I gamma diveristy", title="nosz I")+
  theme(legend.position = "bottom")

plot_16S_y
```

+ Bacterial gamma diversity shows no trend at all with habitat diversity level. 

We proceed to test if regardless of the fact that bacterial gamma diversity **does not** increase with habitat diversity, functional gene gamma diversity is related to bacterial gamma diversity and / or habitat diversity level. 



### nifH

```{r, fig.width = 12, fig.height = 10}

mod_nifh_y <- lm(y_func_div ~ y_effN_bac * Level, data = filter(DIVeffN_y, func_gen == "y_effN_nifh")) 

par(mfrow=c(3,3))
plot(mod_nifh_y)
hist(resid(mod_nifh_y))

shapiro.test(residuals(mod_nifh_y))
ncvTest(mod_nifh_y)
```

The diagnostic plots and tests look fine

#### results
```{r}

anova(mod_nifh_y)

Anova(mod_nifh_y, type=2)


mod_nifh_y %>% tidy %>%  kable

mod_nifh_y %>% glance %>% kable

# predict
PRED <- predict(mod_nifh_y, type = "terms")
Const <- attr(PRED,"constant")

PRED.df <-  data.frame(y = PRED) %>% 
  mutate(y.y_effN_bac.Level = y.y_effN_bac.Level + y.Level + y.y_effN_bac + Const) %>% 
  mutate(y.y_effN_bac = y.y_effN_bac + Const) %>% 
  mutate(y.Level = y.Level + Const) %>% 
  mutate(x = as.vector(DIVeffN_y[DIVeffN_y$func_gen == "y_effN_nifh", ]$y_effN_bac)) %>% 
  mutate(Level = as.vector(DIVeffN_y[DIVeffN_y$func_gen == "y_effN_nifh", ]$Level))

plot_nifh_y <- DIVeffN_y %>% 
  filter(func_gen == "y_effN_nifh") %>% 
  ggplot(aes(x = y_effN_bac, y = y_func_div, colour = as.factor(Level)))+
  geom_point()+
  geom_line(data = PRED.df, aes(x = x, y = y.y_effN_bac.Level), linetype = "dashed")+
  geom_line(data = PRED.df, aes(x = x, y = y.y_effN_bac), colour = "black")+
  theme_bw()+
  labs(x = "bacterial gamma diversity", y = "nosZ I gamma diveristy", title="nosz I")+
  theme(legend.position = "bottom")



plot_nifh_y
```

+ there is no effect of either habitat diversity level or bacterial gamma diversity on the gamma diversity (effective number of types in the big core meta-community) of nifH genes. 


### nosZ I

```{r, fig.width = 12, fig.height = 10}

mod_nosz1_y <- lm(y_func_div ~  y_effN_bac * Level, data = filter(DIVeffN_y, func_gen == "y_effN_nosz1")) 

par(mfrow=c(3,3))
plot(mod_nosz1_y)
hist(resid(mod_nosz1_y))

shapiro.test(residuals(mod_nosz1_y))
ncvTest(mod_nosz1_y)


```

The residuals deviate significantly from a normal distribution (right skewed) and the residuals show heteroscedasticity. We log transform the response variable


```{r, fig.width = 12, fig.height = 10}

mod_nosz1_y_log <- lm(log(y_func_div, 10) ~  y_effN_bac * Level, data = filter(DIVeffN_y, func_gen == "y_effN_nosz1")) 

par(mfrow=c(3,3))
plot(mod_nosz1_y_log)
hist(resid(mod_nosz1_y_log))

shapiro.test(residuals(mod_nosz1_y_log))
ncvTest(mod_nosz1_y_log)

```

It looks not great but better

#### results
```{r}

anova(mod_nosz1_y_log)
Anova(mod_nosz1_y_log, type = 2)


mod_nosz1_y_log %>% tidy %>%  mutate(estimate = 10^estimate) %>%  kable

mod_nosz1_y_log %>% glance %>% kable

# predict
PRED <- predict(mod_nosz1_y_log, type = "terms")
Const <- attr(PRED,"constant")

PRED.df <-  data.frame(y = PRED) %>% 
  mutate(y.y_effN_bac.Level = 10^(y.y_effN_bac.Level + y.Level + y.y_effN_bac + Const)) %>% 
  mutate(y.y_effN_bac = 10^(y.y_effN_bac + Const)) %>% 
  mutate(y.Level = 10^(y.Level + Const)) %>% 
  mutate(x = as.vector(DIVeffN_y[DIVeffN_y$func_gen == "y_effN_nosz1", ]$y_effN_bac)) %>% 
  mutate(Level = as.vector(DIVeffN_y[DIVeffN_y$func_gen == "y_effN_nosz1", ]$Level))

plot_nosz1_y <- DIVeffN_y %>% 
  filter(func_gen == "y_effN_nosz1") %>% 
  ggplot(aes(x = y_effN_bac, y = y_func_div, colour = as.factor(Level)))+
  geom_point()+
  geom_line(data = PRED.df, aes(x = x, y = y.y_effN_bac.Level), linetype = "dashed")+
  geom_line(data = PRED.df, aes(x = x, y = y.y_effN_bac), colour = "black")+
  theme_bw()+
  labs(x = "bacterial gamma diversity", y = "nosZ I gamma diveristy", title="nosZ I")+
  theme(legend.position = "bottom")



plot_nosz1_y
```

+ there is no effect of either habitat diversity level or bacterial gamma diversity on the gamma diversity (effective number of types in the big core metacommunity) of nosZ I genes. 


### nosZ II

```{r, fig.width = 12, fig.height = 10}

mod_nosz2_y <- lm(y_func_div ~  y_effN_bac * Level, data = filter(DIVeffN_y, func_gen == "y_effN_nosz2")) 

par(mfrow=c(3,3))
plot(mod_nosz2_y)
hist(resid(mod_nosz2_y))

shapiro.test(residuals(mod_nosz2_y))
ncvTest(mod_nosz2_y)

```

The diagnostic plots and tests look fine

#### results
```{r}

anova(mod_nosz2_y)

Anova(mod_nosz2_y, type = 2)

mod_nosz2_y %>% tidy %>%  kable

mod_nosz2_y %>% glance %>% kable

# predict
PRED <- predict(mod_nosz2_y, type = "terms")
Const <- attr(PRED,"constant")

PRED.df <-  data.frame(y = PRED) %>% 
  mutate(y.y_effN_bac.Level = y.y_effN_bac.Level + y.Level + y.y_effN_bac + Const) %>% 
  mutate(y.y_effN_bac = y.y_effN_bac + Const) %>% 
  mutate(y.Level = y.Level + Const) %>% 
  mutate(x = as.vector(DIVeffN_y[DIVeffN_y$func_gen == "y_effN_nosz2", ]$y_effN_bac)) %>% 
  mutate(Level = as.vector(DIVeffN_y[DIVeffN_y$func_gen == "y_effN_nosz2", ]$Level))

plot_nosz2_y <- DIVeffN_y %>% 
  filter(func_gen == "y_effN_nosz2") %>% 
  ggplot(aes(x = y_effN_bac, y = y_func_div, colour = as.factor(Level)))+
  geom_point()+
  geom_line(data = PRED.df, aes(x = x, y = y.y_effN_bac.Level), linetype = "dashed")+
  geom_line(data = PRED.df, aes(x = x, y = y.y_effN_bac), colour = "black")+
  theme_bw()+
  labs(x = "bacterial gamma diversity", y = "nosZ II gamma diveristy", title="nosZ II")+
  theme(legend.position = "bottom")


plot_nosz2_y
```

+ nosZ II gamma diversity is significantly and positively correlated to bacterial gamma diversity. 
+ Habitat diversity level has no effect 

## summary Hypothesis I

### functional gene diversity ~ bacterial diversity * sediment

+ figures with predicted best fit lines

```{r, fig.height=10, fig.width=10}
grid.arrange(plot_noszI, plot_noszII, plot_nifh, nrow=2)

```

+ p values

```{r}

data.frame(Factor = rownames(Anova(mod_nifh, type = 2))[1:3],
           nifH = signif(Anova(mod_nifh, type = 2)$`Pr(>F)`[1:3], 2),
           nosZ.I = signif(Anova(mod_nosz1_log, type = 3)$`Pr(>F)`[2:4], 2),
           nosz.II = signif(Anova(mod_nosz2, type = 3)$`Pr(>F)`[2:4], 2)) %>% kable

```

### bacterial gamma diversity ~ habitat diversity

```{r, echo = FALSE}
plot_16S_y
anova(mod_16S)
```

### functional gene gamma diversity ~ bacterial diversity * habitat diversity

+ figures with predicted best fit lines

```{r, fig.height=10, fig.width=10}
grid.arrange(plot_nosz1_y, plot_nosz2_y, plot_nifh_y, nrow=2)

```

```{r}

data.frame(Factor = rownames(Anova(mod_nifh_y, type=2))[1:3],
           nifH = signif(Anova(mod_nifh_y, type=2)$`Pr(>F)`[1:3], 2),
           nosZ.I = signif(Anova(mod_nosz1_y_log, type = 2)$`Pr(>F)`[1:3], 2),
           nosz.II = signif(Anova(mod_nosz2_y, type = 2)$`Pr(>F)`[1:3], 2)) %>% kable

```


# Hypothesis 2

**Diversity of functional community explains process rates better than overall bacterial diversity and or abundance alone**


Here we test the Hypothesis that the two measured processes, denitrification and nitrogen fixation, are - in part - driven by the diversity of the *genetic potential* i.e. the effective number of the respective functional genes. The functional genes that we sequenced encode key components of key proteins necessary for the processes that we measured. A diversity of functional genes hence suggests a diversity of proteins that might operate under different optimal conditions. Hence there is a potential for complementary among the organisms bearing those proteins for using  microniches more efficiently. 

However we do not expect that process rates are governed solely by functional gene diversity and therefore need to include co-variates which we also suspect to influence the measured process rates. These co-variates are 

+ the abundance of functional gene copies in the community (as measure for *genetic potential*)
+ the habitat type (as agglomerative measure for *environmental variables*)
+ the diversity of the whole bacterial community: General bacterial diversity could lead to complementary interactions that favor the bacteria in the functional guilde 


Before we proceed to run the model, we have to address the issue of possible independence of the measurements in each small core within each big core

### test mein effect of MC.ID
```{r}
FuncDat %>% 
  select(N2_fixation, Denitrification, sample) %>% 
  left_join(DIV[1:3]) %>% 
  gather(func, rate, N2_fixation, Denitrification) %>% 
  filter(Level > 1) %>%
  lm(rate ~ func * MC.ID, data=.) %>% 
  anova()
```

process rate ~ func.gen.div + 16Sdiv + func.gen.abund + habitat.type + (1|big core)

There is no significant effect of *big core* on the process rates. Therefore we assume that we can pool the data

### reshape the data

```{r}
FuncDat_sub <- FuncDat %>% 
  select(sample, sediment, N2_fixation, Denitrification, nosZI, nosZII, nifH) %>% 
  left_join(select(DIV, sample, starts_with("effN"))) %>% 
  na.omit()

kable(head(FuncDat_sub))
```


### nifH

```{r, fig.width = 12, fig.height = 10}
mod_proc_nifh <- lm(N2_fixation ~ sediment + nifH + effN_nifh + effN_bac, data = FuncDat_sub)

par(mfrow=c(3,3))
plot(mod_proc_nifh)
hist(resid(mod_proc_nifh))

shapiro.test(residuals(mod_proc_nifh))
ncvTest(mod_proc_nifh)




```

the residuals are left-skewed. We try a sqrt transformation

```{r, fig.width = 12, fig.height = 10}
mod_proc_nifh_sqrt <- lm( sqrt(N2_fixation) ~ sediment + nifH + effN_nifh + effN_bac, data = FuncDat_sub)

par(mfrow=c(3,3))
plot(mod_proc_nifh_sqrt)
hist(resid(mod_proc_nifh_sqrt))

shapiro.test(residuals(mod_proc_nifh_sqrt))
ncvTest(mod_proc_nifh_sqrt)


```

#### results
```{r}
# type I sum of squares testing the interaction
anova(mod_proc_nifh_sqrt)

#type II sum of squares testing the main effects
Anova(mod_proc_nifh_sqrt, type = 2)

mod_proc_nifh_sqrt %>% tidy %>% kable

mod_proc_nifh_sqrt %>% glance %>% kable

# predict
PRED <- predict(mod_proc_nifh_sqrt, type = "terms")
Const <- attr(PRED,"constant")

PRED.df <-  data.frame(PRED) %>%
  mutate(sediment = (sediment + Const)^2) %>% 
  mutate(nifH = (nifH + Const)^2) %>% 
  mutate(effN_nifh = (effN_nifh + Const)^2) %>% 
  mutate(effN_bac = (effN_bac + Const)^2) %>% 
  mutate(sample = as.vector(FuncDat_sub$sample)) %>% 
  gather(predictor, predicted, -sample)

plot_nifh_proc <- 
FuncDat_sub %>% 
  select(match(c("sample", "N2_fixation", unique(PRED.df$predictor)), colnames(.))) %>% 
  mutate(sediment = as.numeric(as.factor(sediment))) %>% 
  gather(predictor, raw, -sample, -N2_fixation) %>% 
  left_join(.,PRED.df) %>% 
  left_join(., select(FuncDat_sub, sample, sediment)) %>%
  mutate(predictor = factor(predictor, levels = c("nifH", "effN_nifh", "effN_bac", "sediment"),
                            labels = c("nifH abundance", "nifH diversity", "bacterial diversity", "sediment"))) %>% 
  ggplot(aes(x = raw, y = N2_fixation, colour = sediment))+
  geom_point()+
  geom_line(aes(x = raw, y = predicted, group = predictor), colour = "black", linetype = "dashed")+
  facet_wrap(~predictor, scales = "free_x")+
  theme_bw()+
  labs(x = "Predictor value", title="nifH - raw data and predicted main effects")+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme(legend.position = "right")

plot_nifh_proc
```

### nosZ I 

```{r, fig.width = 12, fig.height = 10}
mod_proc_nosz1 <- lm(Denitrification ~ sediment + nosZI + effN_nosz1 + effN_bac, data = FuncDat_sub)

par(mfrow=c(3,3))
plot(mod_proc_nosz1)
hist(resid(mod_proc_nosz1))

shapiro.test(residuals(mod_proc_nosz1))
ncvTest(mod_proc_nosz1)


```

The residuals look normal but the variance doesn't look great. Transforming the data doesn't help though (I tried log10 and sqrt) and the variance doesn't look too bad. So we go with it

#### results
```{r}
# type I sum of squares testing the interaction
anova(mod_proc_nosz1)

#type II sum of squares testing the main effects
Anova(mod_proc_nosz1, type = 2)

mod_proc_nosz1 %>% tidy %>% kable

mod_proc_nosz1 %>% glance %>% kable

# predict
PRED <- predict(mod_proc_nosz1, type = "terms")
Const <- attr(PRED,"constant")

PRED.df <-  data.frame(PRED) %>%
  mutate(sediment = sediment + Const) %>% 
  mutate(nosZI = nosZI + Const) %>% 
  mutate(effN_nosz1 = effN_nosz1 + Const) %>% 
  mutate(effN_bac = effN_bac + Const) %>% 
  mutate(sample = as.vector(FuncDat_sub$sample)) %>% 
  gather(predictor, predicted, -sample)

plot_nosz1_proc <- 
FuncDat_sub %>% 
  select(match(c("sample", "Denitrification", unique(PRED.df$predictor)), colnames(.))) %>% 
  mutate(sediment = as.numeric(as.factor(sediment))) %>% 
  gather(predictor, raw, -sample, -Denitrification) %>% 
  left_join(.,PRED.df) %>% 
  left_join(., select(FuncDat_sub, sample, sediment)) %>%
  mutate(predictor = factor(predictor, levels = c("nosZI", "effN_nosz1", "effN_bac", "sediment"),
                            labels = c("nosZ I abundance", "nosZ I diversity", "bacterial diversity", "sediment"))) %>% 
  ggplot(aes(x = raw, y = Denitrification, colour = sediment))+
  geom_point()+
  geom_line(aes(x = raw, y = predicted, group = predictor), colour = "black", linetype = "dashed")+
  facet_wrap(~predictor, scales = "free_x")+
  theme_bw()+
  labs(x = "Predictor value", title="nosZ I - raw data and predicted main effects")+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme(legend.position = "right")

plot_nosz1_proc
```


### nosZ II

```{r, fig.width = 12, fig.height = 10}
mod_proc_nosz2 <- lm(Denitrification ~ sediment + nosZII + effN_nosz2 + effN_bac, data = FuncDat_sub)

par(mfrow=c(3,3))
plot(mod_proc_nosz2)
hist(resid(mod_proc_nosz2))

shapiro.test(residuals(mod_proc_nosz2))
ncvTest(mod_proc_nosz2)

```

#### results
```{r}
# type I sum of squares testing the interaction
anova(mod_proc_nosz2)$`Sum Sq`/sum(anova(mod_proc_nosz2)$`Sum Sq`)*100

#type II sum of squares testing the main effects
Anova(mod_proc_nosz2, type = 2)$`Sum Sq`/sum(Anova(mod_proc_nosz2, type = 2)$`Sum Sq`)*100

mod_proc_nosz2 %>% tidy %>% kable

mod_proc_nosz2 %>% glance %>% kable

# predict
PRED <- predict(mod_proc_nosz2, type = "terms")
Const <- attr(PRED,"constant")

PRED.df <-  data.frame(PRED) %>%
  mutate(sediment = sediment + Const) %>% 
  mutate(nosZII = nosZII + Const) %>% 
  mutate(effN_nosz2 = effN_nosz2 + Const) %>% 
  mutate(effN_bac = effN_bac + Const) %>% 
  mutate(sample = as.vector(FuncDat_sub$sample)) %>% 
  gather(predictor, predicted, -sample)

plot_nosz2_proc <- 
FuncDat_sub %>% 
  select(match(c("sample", "Denitrification", unique(PRED.df$predictor)), colnames(.))) %>% 
  mutate(sediment = as.numeric(as.factor(sediment))) %>% 
  gather(predictor, raw, -sample, -Denitrification) %>% 
  left_join(.,PRED.df) %>% 
  left_join(., select(FuncDat_sub, sample, sediment)) %>%
  mutate(predictor = factor(predictor, levels = c("nosZII", "effN_nosz2", "effN_bac", "sediment"),
                            labels = c("nosZ II abundance", "nosZ II diversity", "bacterial diversity", "sediment"))) %>% 
  ggplot(aes(x = raw, y = Denitrification, colour = sediment))+
  geom_point()+
  geom_line(aes(x = raw, y = predicted, group = predictor), colour = "black", linetype = "dashed")+
  facet_wrap(~predictor, scales = "free_x")+
  theme_bw()+
  labs(x = "Predictor value", title="nosZ II - raw data and predicted main effects")+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme(legend.position = "right")

plot_nosz2_proc
```


## summary Hypothesis 2

```{r, fig.height=10, fig.width=10}
grid.arrange(plot_nosz1_proc, plot_nosz2_proc, plot_nifh_proc, nrow=2)

```

+ p values

```{r}

data.frame(Factor = c("Sediment", "Func. gene abund.", "Func. gene div.", "Bac. div.", "Residuals"),
           nifH.p = signif(Anova(mod_proc_nifh_sqrt, type = 2)$`Pr(>F)`, 2),
           nifH.r = signif(Anova(mod_proc_nifh_sqrt, type = 2)$`Sum Sq`/sum(Anova(mod_proc_nifh_sqrt, type = 2)$`Sum Sq`)*100,2),
           nosZ.I.p = signif(Anova(mod_proc_nosz1, type = 2)$`Pr(>F)`, 2),
           nosZ.I.r = signif(Anova(mod_proc_nosz1, type = 2)$`Sum Sq`/sum(Anova(mod_proc_nosz1, type = 2)$`Sum Sq`)*100,2),
           nosZ.II.p = signif(Anova(mod_proc_nosz2, type = 2)$`Pr(>F)`, 2),
           nosZ.II.r = signif(Anova(mod_proc_nosz2, type = 2)$`Sum Sq`/sum(Anova(mod_proc_nosz2, type = 2)$`Sum Sq`)*100,2)) %>% 
  mutate(nifH.adj = p.adjust(nifH.p, method = "holm")) %>% 
  mutate(nosZ.I.adj = p.adjust(nosZ.I.p, method = "holm")) %>% 
  mutate(nosZ.II.adj = p.adjust(nosZ.II.p, method = "holm")) %>% 
select(Factor,nifH.p,nifH.adj,nifH.r,nosZ.I.p,nosZ.I.adj,nosZ.I.r,nosZ.II.p, nosZ.II.r, nosZ.II.adj) %>% 
  kable
  

```


in the table above:

+ p-values are adjusted with the "holm" adjusting method *within* each model
+ `*.r` is the % total sum of squares for this factor

In conclusion: 

+ all three models explain only a small fraction of the total variance in process rates (15 - 26%)
+ neither of the factors is strongly and consistently related to variation in process rates

# Community composition

We have seen in the previous analysis that the variation in process rates (nitrogen fixation and denitrification respectively) cannot be well explained by either the abundance or the diversity of functional genes nor general bacterial diversity. There is some tendency for the process rates being different in the different sediments but the evidence is not consistent.

An alternative hypothesis would be that process rates are driven by the presence or the combination of a few key species. As we don't have an *a priori* hypothesis abut **which** species could be important, this data analysis is therefore exploratory. 

We use the DESeq2 package to look for differntially OTUs that are 

+ correlated to process rates, controling for sediments. 
+ diffrentialy abundant between the sediments

[Love, M. I., Huber, W., & Anders, S. (2014). Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biology, 15(12), 31.)](http://doi.org/10.1186/s13059-014-0550-8)

for the correlation of abundance with process rates with we test the hypthesis that process rate (PR) and the counts (C) are related as in 

$$PR_{func} = 2^{(aC_{funcgen})} $$
$a$ is the log-fold change of counts with unit process rate. 

We also use the *rlogTransformation* implemented in the DESeq2 package to normalize the OTUtables and cluster the samples. 

## nifH

+ reading in data & data wrangling
```{r}
OTU_nifh <- read.table("OTU_nifh_clean.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

# fix sample names
rownames(OTU_nifh) <- sub("F", "", rownames(OTU_nifh))

# exluding na samples for functional data to match samles in otu table
FuncDat <- FuncDat[!is.na(FuncDat$sample),]
rownames(FuncDat) <- FuncDat$sample

# import phylogenetic tree
TREE_nifh <- read_tree("TREE_nifh_OTUonly.tre")

```

+ we first create a phyloseq object with the otu table, the metadata (the phylogeny is not needed here)
+ we the transform the phyloseq object to a deseq2 object, telling it the experimental design. Here `~sediment + N2_fixation` means that we look for a correlation with N2_fixation, controling for sediment. 
+ we then test the hypothesis using the `DESeq`function. this function does three things: 
    + estimates the size factors (correcting for sequence depth between samples)
    + estimates dispersion (obtains dispersion estimates for Negative Binomial distributed data. See `?estimateDispersions` or the reference above for details. We use the the `parametric`option)
    + GLM fitting and Wald test (tests for significance of coefficients in a Negative Binomial GLM, using previously calculated sizeFactors and dispersion estimates - test for significant $a$ estimates in the formula above)
    + we also set `minReplicatesForReplace` to `Inf`. As we test of a continious variable, we don't have true biological replicates. 
    
### correlation with N2_fixation
```{r}
nifH_ps <- phyloseq(otu_table(OTU_nifh, taxa_are_rows = F),
                    sample_data(FuncDat))

nifH_d2 <- phyloseq_to_deseq2(nifH_ps, ~sediment + N2_fixation)

Sigdiff <- DESeq(nifH_d2, fitType = "parametric", test = "Wald", minReplicatesForReplace = Inf)

res = results(Sigdiff, name = "N2_fixation", cooksCutoff = FALSE)
alpha = 0.05

sigtab <-  res[with(res, order(padj, pvalue)),]
sigtab

plotDispEsts(Sigdiff, main = "dispersion estimations")
plotMA(res, main = "shrunken log2fold changes with unit process rate change")
```

+ if we account for differnce between the sediments, no nifH-OTU is found to be positively (or negatively) correlated with the process rate

### differences between sediments

+ we also check whether we can find OTUs that are signifcantly differnt between the four sediments
+ here, we do have true biological replicates, so we keep `minReplicatesForReplace` activated

becasue it turns out that we find a large number of differntially abundant OTUs between teh sediments we try to focus on the most different cases. Therefore we 

+ test for OTU's that have greater log-fold-changes then |1| (instead of 0) 
+ set a rigurous p-value cutoff of the adjusted p-value of 0.001

```{r}
nifH_d2 <- phyloseq_to_deseq2(nifH_ps, ~sediment)
Sigdiff <- DESeq(nifH_d2, fitType = "parametric", test = "Wald")


# get the results for all contratst and store them list
SedComb <- combn(c("cyano", "ruppia", "sand", "silt"), 2)

alpha = 0.001
LFC = 1

reslist <- list()

for(i in 1:ncol(SedComb)) {
res = results(Sigdiff, contrast = c("sediment", SedComb[,i]), cooksCutoff = FALSE, lfcThreshold = LFC)  
sigtab <- res[res$padj < alpha & !is.na(res$padj),]
reslist[[i]] <- sigtab
}

names(reslist) <- paste(SedComb[1, ], SedComb[2,], sep = "_")

diffOTUs <- unique(unlist(lapply(reslist, rownames)))


OTU.var.st <- getVarianceStabilizedData(Sigdiff)
RL <- rlogTransformation(t(OTU_nifh))
rownames(RL) <- colnames(OTU_nifh)

RL[RL>quantile(c(RL), 0.999)] <- quantile(c(RL), 0.999)

ps <- phyloseq(otu_table(RL[diffOTUs, ], 
                         taxa_are_rows = T),
               sample_data(FuncDat))

plot_heatmap(ps, sample.order = FuncDat[with(FuncDat, order(sediment)), ]$sample,
             sample.label = "sediment", trans = NULL)+
facet_grid(~sediment, scales = "free")


plotMA(res, main = "shrunken log2fold changes with unit process rate change")




res = results(Sigdiff, contrast = c("sediment", "silt", "ruppia"), lfcThreshold =1, cooksCutoff = FALSE)
alpha = 0.01
L2C = 2
sigtab <- res[which(res$padj < alpha), ]
sigtab <- sigtab[which(sigtab$log2FoldChange < L2C), ]

sigtab <-  sigtab[with(sigtab, order(padj)),]
sigtab

plotMA(res)+
abline(1,0,col = "blue")+abline(-1,0, col = "blue")
```


```{r}
#StD <- getVarianceStabilizedData(Sigdiff)
RL <- rlogTransformation(t(OTU_nifh))

rownames(RL) <- colnames(OTU_nifh)

PS <- phyloseq(otu_table(assay(RLp), taxa_are_rows = T),
         phy_tree(TREE_nifh))

UFD <- UniFrac(PS, weighted = TRUE)


RLp <- rlogTransformation(nifH_d2)

NMDS <- metaMDS(t(OTU.var.st))
NMDS <- metaMDS(t(RL))
NMDS <- metaMDS(UFD)

NMDS$points %>% 
  data.frame() %>% 
  rownames_to_column(var = "sample") %>% 
  mutate(sample = as.numeric(sample)) %>% 
  left_join(.,FuncDat) %>% 
  ggplot(aes(x = MDS1, y =MDS2, colour = sediment))+
  geom_point()+
  theme_bw()+
  ggtitle("nifH gene composition - bray-curtis - rlogTransformed")

```


```{r}
dd <- plotCounts(Sigdiff, gene=c(row.names(sigtab)[1]), intgroup="N2_fixation",
returnData=TRUE)
dd$OTU <- row.names(sigtab)[1]

for (i in row.names(sigtab)[-1]) {
ddt <- plotCounts(Sigdiff, gene=i, intgroup="N2_fixation",
returnData=TRUE)

ddt$OTU <- i

dd <- rbind(dd,ddt)
}

left_join(dd, FuncDat) %>% 
ggplot(aes(y=N2_fixation, x=count, colour = sediment)) +
  geom_point() +
  facet_wrap(~OTU)+
  theme_bw()+
  scale_x_log10()
```




```{r}
OTU_nifh <- decostand(OTU_nifh, "total", 1)*100
```

max and median relative abundances

```{r, fig.show='hold', fig.width = 5}

apply(OTU_nifh, 2, max) %>% 
  data.frame() %>% 
  ggplot( aes(x = `.`))+
  geom_histogram(binwidth = 0.5, colour = "#c9c9c9", fill = "#98bdd3") +
  scale_y_sqrt(breaks=seq(0,40,2)^2)+
  scale_x_continuous(breaks=seq(0,10,0.5))+
  theme_bw()+
  ggtitle("maximum relative abundance")

apply(OTU_nifh, 2, median) %>% 
  data.frame() %>% 
  ggplot( aes(x = `.`))+
  geom_histogram(binwidth = 0.005, colour = "#c9c9c9", fill = "#98bdd3") +
  scale_y_sqrt(breaks=seq(0,40,2)^2)+
  scale_x_continuous()+
  theme_bw()+
  ggtitle("median relative abundance")

apply(OTU_nifh, 2, function(x){sd(x)/mean(x)}) %>% 
  data.frame() %>% 
  ggplot( aes(x = `.`))+
  geom_histogram(binwidth = 0.5, colour = "#c9c9c9", fill = "#98bdd3") +
  scale_y_sqrt(breaks=seq(0,40,2)^2)+
  scale_x_continuous()+
  theme_bw()+
  ggtitle("cv")
```

The distribution of maximum and median relative abundance suggests that we can start 
