---
title: "Diversity estimates"
output:
  html_notebook:
    toc: yes
    toc_depth: 4
    toc_float: yes
  html_document:
    toc: yes
    toc_depth: 4
    toc_float: yes
---

```{r, setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, highlight = TRUE)

# change 'var' to something else to invalidate all cached chunks (i.e. force to recompute them) 

Var <- "1"
knitr::opts_chunk$set(cache.whatever = quote(Var))
```

```{r, echo = F}
library(knitr)
library(tidyverse)
library(entropart)
library(ape)
library(vegan)
library(phyloseq)
library(GGally)
library(parallel)
library(RColorBrewer)
```

# Experimental design:

![experimental set-up](Experimental_layout.jpg)

In the experimental design, each diversity treatments consist of a meta-community ("big core") of four separate sediment communities ("small core") that correspond to the habitats.

For habitat level 1, each **metacommunity** is replicated 4 times for each habitat (resulting in 16 meta-communities in total)

habitat level 2 to 4, consist of four replicated meta-communities each, and each meta-community contains 2x2 [level 2], (2x1 + 1x2) [level 3] and 4x1 [level 4] habitat communities ("small cores"), respectively. 

There are two missing samples:

+ Level 3 - replicate A - habitat cyano
+ Level 3 - replicate D - habitat silt


We use data only from the summer (excluding spring and autumn).

### Import metadata 

```{r, cache = TRUE}
meta <- read.table("meta_clean.txt") 
```


#nifH

Read in the relevant files:
```{r, cache = TRUE}
OTU_nifh <- read.table("OTU_nifh_clean.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
TREE_nifh <- read.tree("TREE_nifh_um_oT.tre")
```

+ remove "F" preceding sample number in OTU table


```{r, cache = TRUE}
dimnames(OTU_nifh)[[1]] <- sub("F", "", dimnames(OTU_nifh)[[1]])
OTU_nifh <- OTU_nifh[rownames(OTU_nifh) %in% meta$sample, ]
```

### Sampling depth 

The distribution of sampling depth among all pooled samples is:
```{r, cache = TRUE}
barplot(sort(rowSums(OTU_nifh), decreasing = TRUE), col = "white")
```


### Rarefaction

Here we plot the sample rarefaction curves to check if it reached an asymptote for all samples. Note that we plot the effective number of species (q = 1) against sampling depth, not the richness.

```{r, cache = TRUE, eval = TRUE}
# wrapper for `seq` for quadratic spaced sequences
qseq <- function(from, to, length.out) {seq(sqrt(from), sqrt(to), length.out = length.out)^2}

nifH_rar <- split(OTU_nifh, 1:nrow(OTU_nifh))

RareDF_nifh <- mclapply(nifH_rar, function(x) {
    m <-  round(qseq(1, ifelse(sum(x) < 15000, sum(x), 15000), length.out = 30))
    rar <- iNEXT:::Dqhat.Ind(x, q = 1, m)
    out <- data.frame(m = m, qD = rar)
    return(out)})

names(RareDF_nifh) <- row.names(OTU_nifh)

RareDF_nifh_df <- do.call("rbind",RareDF_nifh )

RareDF_nifh_df$sample <- as.numeric(gsub("(\\d+).+", "\\1", row.names(RareDF_nifh_df)))

RareDF_nifh_df %>% 
  left_join(meta) %>% 
  na.omit() %>% 
  ggplot(aes(x = m, y = qD, colour = sediment, group = sample))+
  geom_line() +
  facet_wrap(~Level, labeller = label_wrap_gen(multi_line=FALSE))+
  theme_bw()+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))+
  labs(y = "diversity (q = 1)", x = "sampling depth", title = "rarefaction")
```

## Alpha diversity

### Richness and effective number of species

```{r, cache = TRUE}

### Richness ###
CommunityEntropies <- apply(OTU_nifh, 1, bcTsallis, q = 0, CheckArguments=FALSE, Correction = "ChaoWangJost")

DIV_S_nifh <- data.frame(sample = as.numeric(names(CommunityEntropies)), 
                       S_nifh = CommunityEntropies, 
                       row.names=NULL, stringsAsFactors = F)


#### Effective number of species ###
CommunityEntropies <- apply(OTU_nifh, 1, bcTsallis, q = 1, CheckArguments=FALSE, Correction = "Marcon")
CommunityDiversity <- exp(CommunityEntropies)

DIV_effN_nifh <- data.frame(sample = as.numeric(names(CommunityDiversity)), 
                       effN_nifh = CommunityDiversity, 
                       row.names=NULL, stringsAsFactors = F)

```

## Phylogenetic diversity

```{r, cache = TRUE}
ppTree <- Preprocess.Tree(multi2di(TREE_nifh))

CommunityEntropies_phylo <- apply(t(OTU_nifh), 2, bcPhyloEntropy, Tree = ppTree, q = 1, CheckArguments=FALSE, Correction = "Marcon")

CommunityDiversity_phylo <- unlist( lapply( CommunityEntropies_phylo, function(x) exp(x$Total)))

DIV_effN_phy_nifh <- data.frame(sample = as.numeric(names(CommunityDiversity_phylo)), 
                           phyloDiv_nifh = CommunityDiversity_phylo, row.names=NULL)

DIV_nifh <- 
  left_join(DIV_S_nifh, DIV_effN_nifh) %>% 
  left_join(DIV_effN_phy_nifh)
  

ggscatmat(DIV_nifh[,2:4])+
  theme_bw()
```


## Gamma diversity

+ First we need to create a unique identifier for each *big core* (see schema of experimental design).

```{r, cache = TRUE}
meta$MC.ID <- paste(meta$Level, meta$replicate, sep="_")

meta[meta$Level == 1, ]$MC.ID <- paste(meta[meta$Level == 1, ]$MC.ID, substr(meta[meta$Level == 1, ]$sediment, 1, 2), sep = "_")
```


In this step we calculate gamma diversit at the level of the big cores

+ For the Level 1 cores we don't need to calculate gamma diversity. We only have one pooled measurement per large core as we only have one sediment type in each level 1 core.

+ For level 2 to 4, we calculate the gamma richness, effective number of species and phylogenetic diversity as for the small cores but for the metacommunity weighted by the relative abundance of each small core (only relevant for level 3).

+ We calculate gamma diversity for all cores were we have at least 2 samples. However 1 core, from level 3 has only a single sample left (the two other were excluded). Here we cannot calculate gamm diversity. 

```{r}
meta %>% filter(Level > 1) %>% group_by(MC.ID) %>% filter(n() == 1) 
```


```{r, cache = TRUE}

y_DIV_nifh <- data.frame(MC.ID = unique(meta[meta$Level != 1 & meta$MC.ID != "3_D", ]$MC.ID),
                    y_S_nifh = NA,
                    y_effN_nifh = NA,
                    y_phyloDiv_nifh = NA)


for(i in y_DIV_nifh$MC.ID) {
  
  # Extract OTU data from metacommunity i
  OTU_temp <-  OTU_nifh[ which( rownames( OTU_nifh) %in% meta[ meta$MC.ID == i ,]$sample), ]
  
  # Metacommunity object with weights
  MC_temp <- MetaCommunity(t(OTU_temp), meta[meta$sample %in% as.numeric(rownames(OTU_temp)),]$weight)
  
  # Calculate phylogenetic Alpha diveristy of order q = 1, with communities weighted by there relative proportion
  y_DIV_nifh[y_DIV_nifh$MC.ID == i,]$y_S_nifh <- GammaDiversity(MC_temp, q = 0, Correction = "ChaoWangJost")
  y_DIV_nifh[y_DIV_nifh$MC.ID == i,]$y_effN_nifh <- GammaDiversity(MC_temp, q = 1, Correction = "Marcon")
  y_DIV_nifh[y_DIV_nifh$MC.ID == i,]$y_phyloDiv_nifh <- GammaDiversity(MC_temp, ppTree, q = 1, Correction = "Marcon")
  

}


y_DIV_nifh <- 
y_DIV_nifh %>% 
  left_join( meta, .) %>% 
  #na.omit() %>% 
  select(Level, sample, MC.ID, sediment, y_S_nifh, y_effN_nifh, y_phyloDiv_nifh) %>% 
  left_join(DIV_nifh) %>% 
  mutate(y_S_nifh = ifelse(Level == 1, S_nifh, y_S_nifh)) %>% 
  mutate(y_effN_nifh = ifelse(Level == 1, effN_nifh, y_effN_nifh)) %>% 
  mutate(y_phyloDiv_nifh = ifelse(Level == 1, phyloDiv_nifh, y_phyloDiv_nifh))

write.table(y_DIV_nifh, "DIV_nifh.txt")

```
## potential Gamma diversity

```{r}

meta1 <- filter(meta, Level == 1, replicate != "A")

Comb_L <- split(meta1$MC.ID, meta1$sediment)

Lcomb <- list(combn(4,2), combn(4,3), combn(4,4))

Comb_L_all <- list()

for (l in seq_along(Lcomb)){
  for (k in seq_len(ncol(Lcomb[[l]]))){
    temp <- expand.grid(Comb_L[Lcomb[[l]][,k]], stringsAsFactors = FALSE)
    name <- paste(colnames(temp), collapse = "_")
    Comb_L_all[[name]] <- temp
  }
}


y_nifh_pot <- data.frame(sed_comb = rep(names(Comb_L_all), unlist(lapply(Comb_L_all, nrow))),
                          comb_n = unlist(lapply(lapply(Comb_L_all, nrow), function(x) 1:x)),
                          sed_n = NA,
                          y_S = NA,
                          y_effN = NA)


for(i in names(Comb_L_all)){
  for(k in seq_len(nrow(Comb_L_all[[i]]))){

  # Extract OTU data from metacommunity i
    
  temp_semp <- meta[meta$MC.ID %in% Comb_L_all[[i]][k,],]$sample
  
  OTU_temp <-  OTU_nifh[ which( rownames( OTU_nifh) %in% temp_semp), ]
  
# Metacommunity object with weights
MC_temp <- MetaCommunity(t(OTU_temp))
  
  # Calculate phylogenetic Alpha diveristy of order q = 1, with communities weighted by there relative proportion
  y_nifh_pot[y_nifh_pot$sed_comb == i & y_nifh_pot$comb_n == k ,]$y_S <- 
    GammaDiversity(MC_temp, q = 0, Correction = "ChaoWangJost")
  y_nifh_pot[y_nifh_pot$sed_comb == i & y_nifh_pot$comb_n == k ,]$y_effN <- 
    GammaDiversity(MC_temp, q = 1, Correction = "Marcon")
  y_nifh_pot[y_nifh_pot$sed_comb == i & y_nifh_pot$comb_n == k ,]$sed_n <- length(temp_semp)
    
}}


y_nifh_pot <- 
y_DIV_nifh %>% 
  group_by(MC.ID) %>% 
  mutate(sed_comb = paste(sediment, collapse = "_"),
         sed_n = n(), comb_n = NA) %>% 
  ungroup() %>% 
  rename( y_S = y_S_nifh , y_effN = y_effN_nifh) %>% 
  select(y_S, y_effN, sed_comb, sed_n, comb_n) %>% 
  distinct(.keep_all = FALSE) %>% 
  rbind(y_nifh_pot) %>% 
  group_by(sed_n) %>% 
  mutate(weight = n()) %>% 
  mutate(Obs = ifelse(is.na(comb_n), "observed", "potential"))

y_nifh_pot$sed_comb <- factor(y_nifh_pot$sed_comb, 
                               levels = y_nifh_pot[with(y_nifh_pot, order(sed_n, sed_comb)),]$sed_comb)

y_nifh_pot_plot <- ggplot(y_nifh_pot, aes(x = as.factor(sed_n), y = y_effN, fill = sed_comb, shape = Obs, alpha = Obs))+
  geom_point(size = 3, colour = "black",
             position = position_dodge(width = 0.7))+
  geom_smooth(method = "lm", se = F, aes(group = 1, weight = 1/weight), colour = "black", size = 0.3)+
  scale_fill_manual(values = c("cyan4", "forestgreen", "tan2", "brown4",
                               brewer.pal(6, "Set2"),
                               brewer.pal(4, "Dark2"),
                               brewer.pal(1, "Paired")))+
  scale_shape_manual(values = c(22,21))+
  scale_alpha_manual(values = c(1,0.6))+
  guides(fill = guide_legend(override.aes = list(shape = 21)))+
  labs(x = "Habitat richness", y = "Effective number of species", title = "nifH")+
  theme_bw()

y_nifh_pot_plot
```

# 16S

Read in the relevant files:

```{r, cache = TRUE}
OTU_bac <- read.table("16S_analysis/OTU_bac_clean.txt", header = TRUE, stringsAsFactors = FALSE)
TREE_bac <- read.tree("16S_analysis/Tree_bac_um_oT.tre")
```

+ Remove "F" preceding sample number in OTU table.
+ Subset from summer sample.

```{r, cache = TRUE}
dimnames(OTU_bac)[[1]] <- sub("C", "", dimnames(OTU_bac)[[1]])
OTU_bac <- OTU_bac[row.names(OTU_bac) %in% meta$sample,]
```

### Sampling depth 

The distribution of sampling depth among all pooled samples is:
```{r, cache = TRUE, echo=FALSE}
barplot(sort(rowSums(OTU_bac), decreasing = TRUE), col = "white")
```


### Rarefaction

```{r, cache = TRUE, eval = TRUE}
bac_rar <-  split(OTU_bac, 1:nrow(OTU_bac))

RareDF_bac <- mclapply(bac_rar, function(x) {
    m <-  round(qseq(1, ifelse(sum(x) < 30000, sum(x), 30000), length.out = 30))
    rar <- iNEXT:::Dqhat.Ind(x, q = 1, m)
    out <- data.frame(m = m, qD = rar)
    return(out)})

names(RareDF_bac) <- row.names(OTU_bac)

RareDF_bac_df <- do.call("rbind", RareDF_bac)

RareDF_bac_df$sample <- as.numeric(gsub("(\\d+).+", "\\1", row.names(RareDF_bac_df)))

RareDF_bac_df %>% 
  left_join(meta) %>% 
  na.omit() %>% 
  ggplot(aes(x = m, y = qD, colour = sediment, group = sample))+
  facet_wrap(~Level, labeller = label_wrap_gen(multi_line=FALSE))+
  geom_line() +
  theme_bw()+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))+
  labs(y = "diversity (q = 1)", x = "sampling depth", title = "rarefaction")

```

## Alpha diversity

## Richness and effective number of species

```{r, cache = TRUE}

### Richness ###
CommunityEntropies <- apply(OTU_bac, 1, bcTsallis, q = 0, CheckArguments=FALSE, Correction = "ChaoWangJost")

DIV_S_bac <- data.frame(sample = as.numeric(names(CommunityEntropies)), 
                       S_bac = CommunityEntropies, 
                       row.names=NULL, stringsAsFactors = F)


#### Effective number of species ###
cl <- makeCluster(4)

CommunityEntropies <- apply(OTU_bac, 1, bcTsallis, q = 1, CheckArguments=FALSE, Correction = "Marcon")
CommunityDiversity <- exp(CommunityEntropies)

DIV_effN_bac <- data.frame(sample = as.numeric(names(CommunityDiversity)), 
                       effN_bac = CommunityDiversity, 
                       row.names=NULL, stringsAsFactors = F)

```

## Phylogenetic diversity

```{r, cache = TRUE}
ppTree <- Preprocess.Tree(multi2di(TREE_bac))

CommunityEntropies_phylo <- apply(OTU_bac, 1, bcPhyloEntropy, Tree = ppTree, q = 1, CheckArguments=FALSE, Correction = "Marcon")

CommunityDiversity_phylo <- unlist( lapply( CommunityEntropies_phylo, function(x) exp(x$Total)))

DIV_effN_phy_bac <- data.frame(sample = as.numeric(names(CommunityDiversity_phylo)), 
                           phyloDiv_bac = CommunityDiversity_phylo, row.names=NULL)

DIV_bac <- 
  left_join(DIV_S_bac, DIV_effN_bac) %>% 
  left_join(DIV_effN_phy_bac)
  

ggscatmat(DIV_bac[,2:4])+
  theme_bw()
```


## Gamma diversity

In this step we pool the samples from the complete small cores. 

+ For the Level 1 cores we don't need to calculate gamma diversity. We only have one pooled measurement per large core as we only have one sediment type in each level 1 core.

+ For level 2 to 4, we calculate the gamma richness, effective number of species and phylogenetic diversity as for the small cores but for the metacommunity weighted by the relative abundance of each small core.

```{r, cache = TRUE}


y_DIV_bac <- data.frame(MC.ID = unique(meta[meta$Level != 1 & meta$MC.ID != "3_D", ]$MC.ID),
                    y_S_bac = NA,
                    y_effN_bac = NA,
                    y_phyloDiv_bac = NA)


for(i in y_DIV_bac$MC.ID) {
  # Extract OTU data from metacommunity i
  OTU_temp <-  OTU_bac[ which( rownames( OTU_bac) %in% meta[ meta$MC.ID == i ,]$sample), ]
  
  # Metacommunity object with weights
  MC_temp <- MetaCommunity(t(OTU_temp), meta[meta$sample %in% as.numeric(rownames(OTU_temp)),]$weight)
  
  # Calculate phylogenetic Alpha diveristy of order q = 1, with communities weighted by there relative proportion
  y_DIV_bac[y_DIV_bac$MC.ID == i,]$y_S_bac <- GammaDiversity(MC_temp, q = 0, Correction = "ChaoWangJost")
  y_DIV_bac[y_DIV_bac$MC.ID == i,]$y_effN_bac <- GammaDiversity(MC_temp, q = 1, Correction = "Marcon")
  y_DIV_bac[y_DIV_bac$MC.ID == i,]$y_phyloDiv_bac <- GammaDiversity(MC_temp, ppTree, q = 1, Correction = "Marcon")
  

}

y_DIV_bac <- 
y_DIV_bac %>% 
  left_join( meta, .) %>% 
  select(Level, sample, MC.ID, sediment, y_S_bac, y_effN_bac, y_phyloDiv_bac) %>% 
  left_join(DIV_bac) %>% 
  mutate(y_S_bac = ifelse(Level == 1, S_bac, y_S_bac)) %>% 
  mutate(y_effN_bac = ifelse(Level == 1, effN_bac, y_effN_bac)) %>% 
  mutate(y_phyloDiv_bac = ifelse(Level == 1, phyloDiv_bac, y_phyloDiv_bac))

write.table(y_DIV_bac, "DIV_bac.txt")

```
## potential Gamma diversity

```{r}

meta1 <- filter(meta, Level == 1, replicate != "A")

Comb_L <- split(meta1$MC.ID, meta1$sediment)

Lcomb <- list(combn(4,2), combn(4,3), combn(4,4))

Comb_L_all <- list()

for (l in seq_along(Lcomb)){
  for (k in seq_len(ncol(Lcomb[[l]]))){
    temp <- expand.grid(Comb_L[Lcomb[[l]][,k]], stringsAsFactors = FALSE)
    name <- paste(colnames(temp), collapse = "_")
    Comb_L_all[[name]] <- temp
  }
}


y_bac_pot <- data.frame(sed_comb = rep(names(Comb_L_all), unlist(lapply(Comb_L_all, nrow))),
                          comb_n = unlist(lapply(lapply(Comb_L_all, nrow), function(x) 1:x)),
                          sed_n = NA,
                          y_S = NA,
                          y_effN = NA)


for(i in names(Comb_L_all)){
  for(k in seq_len(nrow(Comb_L_all[[i]]))){

  # Extract OTU data from metacommunity i
    
  temp_semp <- meta[meta$MC.ID %in% Comb_L_all[[i]][k,],]$sample
  
  OTU_temp <-  OTU_bac[ which( rownames( OTU_bac) %in% temp_semp), ]
  
# Metacommunity object with weights
MC_temp <- MetaCommunity(t(OTU_temp))
  
  # Calculate phylogenetic Alpha diveristy of order q = 1, with communities weighted by there relative proportion
  y_bac_pot[y_bac_pot$sed_comb == i & y_bac_pot$comb_n == k ,]$y_S <- 
    GammaDiversity(MC_temp, q = 0, Correction = "ChaoWangJost")
  y_bac_pot[y_bac_pot$sed_comb == i & y_bac_pot$comb_n == k ,]$y_effN <- 
    GammaDiversity(MC_temp, q = 1, Correction = "Marcon")
  y_bac_pot[y_bac_pot$sed_comb == i & y_bac_pot$comb_n == k ,]$sed_n <- length(temp_semp)
    
}}


y_bac_pot <- 
y_DIV_bac %>% 
  group_by(MC.ID) %>% 
  mutate(sed_comb = paste(sediment, collapse = "_"),
         sed_n = n(), comb_n = NA) %>% 
  ungroup() %>% 
  rename( y_S = y_S_bac , y_effN = y_effN_bac) %>% 
  select(y_S, y_effN, sed_comb, sed_n, comb_n) %>% 
  distinct(.keep_all = FALSE) %>% 
  rbind(y_bac_pot) %>% 
  group_by(sed_n) %>% 
  mutate(weight = n()) %>% 
  mutate(Obs = ifelse(is.na(comb_n), "observed", "potential"))

y_bac_pot$sed_comb <- factor(y_bac_pot$sed_comb, 
                               levels = y_bac_pot[with(y_bac_pot, order(sed_n, sed_comb)),]$sed_comb)

y_bac_pot_plot <- ggplot(y_bac_pot, aes(x = as.factor(sed_n), y = y_effN, fill = sed_comb, shape = Obs, alpha = Obs))+
  geom_point(size = 3, colour = "black",
             position = position_dodge(width = 0.7))+
  geom_smooth(method = "lm", se = F, aes(group = 1, weight = 1/weight), colour = "black", size = 0.3)+
  scale_fill_manual(values = c("cyan4", "forestgreen", "tan2", "brown4",
                               brewer.pal(6, "Set2"),
                               brewer.pal(4, "Dark2"),
                               brewer.pal(1, "Paired")))+
  scale_shape_manual(values = c(22,21))+
  scale_alpha_manual(values = c(1,0.6))+
  guides(fill = guide_legend(override.aes = list(shape = 21)))+
  labs(x = "Habitat richness", y = "Effective number of species", title = "16S")+
  theme_bw()

y_bac_pot_plot
```

# nosZ

### Preprocessing nosz files

Read in the relevant files:

+ otu_table noszI
```{r, cache = TRUE}
OTU_nosz1 <- read.table("nosz/nosZI_OTUtable_0.97_FR.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
dimnames(OTU_nosz1) <-  list(OTU_nosz1$OTU_ID, colnames(OTU_nosz1))
OTU_nosz1 <- t(as.matrix(OTU_nosz1[,-1 ]))
dimnames(OTU_nosz1)[[1]] <- sub("I", "", dimnames(OTU_nosz1)[[1]])
OTU_nosz1 <- OTU_nosz1[rownames(OTU_nosz1) %in% meta$sample, ]
```


+ otu_table noszII
```{r, cache = TRUE}
OTU_nosz2 <- read.table("nosz/nosZII_OTUtable_0.97_FR.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
dimnames(OTU_nosz2) <-  list(OTU_nosz2$OTU_ID, colnames(OTU_nosz2))
OTU_nosz2 <- t(as.matrix(OTU_nosz2[,-1 ]))
dimnames(OTU_nosz2)[[1]] <- sub("II", "", dimnames(OTU_nosz2)[[1]])
OTU_nosz2 <- OTU_nosz2[rownames(OTU_nosz2) %in% meta$sample, ]

```

+ Phylogeny noszI & II
```{r, cache = TRUE}
TREE_nosz <- read.tree("nosz/nosZ_refs_reps_aligned.tree")

phlyo_nosz <- phyloseq(otu_table( cbind(OTU_nosz1, OTU_nosz2), taxa_are_rows = FALSE),
                      phy_tree(TREE_nosz),
                      tax_table(matrix(c(rep("nozI", ncol(OTU_nosz1)), rep("noszII", ncol(OTU_nosz2))),
                                       dimnames = list(c(colnames(OTU_nosz1), colnames(OTU_nosz2)), "nosz"))))

plot_tree(phlyo_nosz, color = "nosz")

```

+ Making tree ultrametric

```{r, cache = TRUE}

write.tree(TREE_nosz, "nosz/nosZ_refs_reps_aligned_FR.tree")

system("/Applications/PATHd8/PATHd8 nosz/nosZ_refs_reps_aligned_FR.tree nosz/nosZ_um.tree")
system("grep '^d8 tree' nosz/nosZ_um.tree > nosz/nosZ_um_oT.tree")
```

```{r, cache = TRUE}
TREE_nosz <- read.tree("nosz/nosZ_um_oT.tree")

phlyo_nosz <- phyloseq(otu_table( cbind(OTU_nosz1, OTU_nosz2), taxa_are_rows = FALSE),
                      phy_tree(TREE_nosz),
                      tax_table(matrix(c(rep("nozI", ncol(OTU_nosz1)), rep("noszII", ncol(OTU_nosz2))),
                                       dimnames = list(c(colnames(OTU_nosz1), colnames(OTU_nosz2)), "nosz"))))

plot_tree(phlyo_nosz, color = "nosz")
  

```

+ Split Tree into two trees for noszI and noszII.
+ Delete all non-OTU sequences during the process.
+ Keep one OTU of noszI and noszII in the respective other tree, to keep root edge.

```{r, cache = TRUE}
TREE_nosz1 <- drop.tip(TREE_nosz, TREE_nosz$tip.label[! TREE_nosz$tip.label %in% c(colnames(OTU_nosz1),colnames(OTU_nosz2)[1])])

phlyo_noszI <- phyloseq(otu_table( cbind(OTU_nosz1, OTU_nosz2), taxa_are_rows = FALSE),
                      phy_tree(TREE_nosz1),
                      tax_table(matrix(c(rep("nozI", ncol(OTU_nosz1)), rep("noszII", ncol(OTU_nosz2))),
                                       dimnames = list(c(colnames(OTU_nosz1), colnames(OTU_nosz2)), "nosz"))))

plot_tree(phlyo_noszI, color = "nosz")
```

```{r, cache = TRUE}
TREE_nosz2 <- drop.tip(TREE_nosz, TREE_nosz$tip.label[! TREE_nosz$tip.label %in% c(colnames(OTU_nosz2),colnames(OTU_nosz1)[1])])

phlyo_noszII <- phyloseq(otu_table( cbind(OTU_nosz1, OTU_nosz2), taxa_are_rows = FALSE),
                      phy_tree(TREE_nosz2),
                      tax_table(matrix(c(rep("nozI", ncol(OTU_nosz1)), rep("noszII", ncol(OTU_nosz2))),
                                       dimnames = list(c(colnames(OTU_nosz1), colnames(OTU_nosz2)), "nosz"))))

plot_tree(phlyo_noszII, color = "nosz")
```

## nosZ I

### Sampling depth 

The distribution of sampling depth among all pooled samples is:
```{r, cache = TRUE}
barplot(sort(rowSums(OTU_nosz1), decreasing = TRUE), col = "white")
```


### Rarefaction

```{r, cache = TRUE}
nosz1_rar <-  split(OTU_nosz1, 1:nrow(OTU_nosz2))

RareDF_nosz1 <- mclapply(nosz1_rar, function(x) {
    m <-  round(qseq(1, ifelse(sum(x) < 6000, sum(x), 6000), length.out = 30))
    rar <- iNEXT:::Dqhat.Ind(x, q = 1, m)
    out <- data.frame(m = m, qD = rar)
    return(out)})

names(RareDF_nosz1) <- row.names(OTU_nosz1)

RareDF_nosz1_df <- do.call("rbind", RareDF_nosz1)

RareDF_nosz1_df$sample <- as.numeric(gsub("(\\d+).+", "\\1", row.names(RareDF_nosz1_df)))

RareDF_nosz1_df %>% 
  left_join(meta) %>% 
  na.omit() %>% 
  ggplot(aes(x = m, y = qD, colour = sediment, group = sample))+
  facet_wrap(~Level, labeller = label_wrap_gen(multi_line=FALSE))+
  geom_line() +
  theme_bw()+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))+
  labs(y = "diversity (q = 1)", x = "sampling depth", title = "rarefaction")
```
### Alpha diversity

### Richness and effective number of species

```{r, cache = TRUE}

### Richness ###
CommunityEntropies <- apply(OTU_nosz1, 1, bcTsallis, q = 0, CheckArguments=FALSE, Correction = "ChaoWangJost")

DIV_S_nosz1 <- data.frame(sample = as.numeric(names(CommunityEntropies)), 
                       S_nosz1 = CommunityEntropies, 
                       row.names=NULL, stringsAsFactors = F)


#### Effective number of species ###
cl <- makeCluster(2)

CommunityEntropies <- apply(OTU_nosz1, 1, bcTsallis, q = 1, CheckArguments=FALSE, Correction = "Marcon")
CommunityDiversity <- exp(CommunityEntropies)

DIV_effN_nosz1 <- data.frame(sample = as.numeric(names(CommunityDiversity)), 
                       effN_nosz1 = CommunityDiversity, 
                       row.names=NULL, stringsAsFactors = F)

```

### Phylogenetic diversity

```{r, cache = TRUE}
ppTree <- Preprocess.Tree(multi2di(TREE_nosz1))

CommunityEntropies_phylo <- apply(OTU_nosz1, 1, bcPhyloEntropy, Tree = ppTree, q = 1, CheckArguments=FALSE, Correction = "Marcon")

CommunityDiversity_phylo <- unlist( lapply( CommunityEntropies_phylo, function(x) exp(x$Total)))

DIV_effN_phy_nosz1 <- data.frame(sample = as.numeric(names(CommunityDiversity_phylo)), 
                           phyloDiv_nosz1 = CommunityDiversity_phylo, row.names=NULL)

DIV_nosz1 <- 
  left_join(DIV_S_nosz1, DIV_effN_nosz1) %>% 
  left_join(DIV_effN_phy_nosz1)
  

ggscatmat(DIV_nosz1[,2:4])+
  theme_bw()
```


## Gamma diversity

In this step we pool the samples from the complete small cores.

+ For the Level 1 cores we don't need to calculate gamma diversity. We only have one pooled measurement per large core as we only have one sediment type in each level 1 core.

+ For level 2 to 4, we calculate the gamma richness, effective number of species and phylogenetic diversity as for the small cores but for the metacommunity weighted by the relative abundance of each small core.

```{r, cache = TRUE}


y_DIV_nosz1 <- data.frame(MC.ID = unique(meta[meta$Level != 1 & meta$MC.ID != "3_D", ]$MC.ID),
                    y_S_nosz1 = NA,
                    y_effN_nosz1 = NA,
                    y_phyloDiv_nosz1 = NA)


for(i in y_DIV_nosz1$MC.ID) {
  # Extract OTU data from metacommunity i
  OTU_temp <-  OTU_nosz1[ which( rownames( OTU_nosz1) %in% meta[ meta$MC.ID == i ,]$sample), ]
  
  # Metacommunity object with weights
  MC_temp <- MetaCommunity(t(OTU_temp), meta[meta$sample %in% as.numeric(rownames(OTU_temp)),]$weight)
  
  # Calculate phylogenetic Alpha diveristy of order q = 1, with communities weighted by there relative proportion
  y_DIV_nosz1[y_DIV_nosz1$MC.ID == i,]$y_S_nosz1 <- GammaDiversity(MC_temp, q = 0, Correction = "ChaoWangJost")
  y_DIV_nosz1[y_DIV_nosz1$MC.ID == i,]$y_effN_nosz1 <- GammaDiversity(MC_temp, q = 1, Correction = "Marcon")
  y_DIV_nosz1[y_DIV_nosz1$MC.ID == i,]$y_phyloDiv_nosz1 <- GammaDiversity(MC_temp, ppTree, q = 1, Correction = "Marcon")
  

}

y_DIV_nosz1 <- 
y_DIV_nosz1 %>% 
  left_join( meta, .) %>% 
  select(Level, sample, MC.ID, sediment, y_S_nosz1, y_effN_nosz1, y_phyloDiv_nosz1) %>% 
  left_join(DIV_nosz1) %>% 
  mutate(y_S_nosz1 = ifelse(Level == 1, S_nosz1, y_S_nosz1)) %>% 
  mutate(y_effN_nosz1 = ifelse(Level == 1, effN_nosz1, y_effN_nosz1)) %>% 
  mutate(y_phyloDiv_nosz1 = ifelse(Level == 1, phyloDiv_nosz1, y_phyloDiv_nosz1))

write.table(y_DIV_nosz1, "DIV_nosz1.txt")

```

## potential Gamma diversity

```{r}

meta1 <- filter(meta, Level == 1, replicate != "A")

Comb_L <- split(meta1$MC.ID, meta1$sediment)

Lcomb <- list(combn(4,2), combn(4,3), combn(4,4))

Comb_L_all <- list()

for (l in seq_along(Lcomb)){
  for (k in seq_len(ncol(Lcomb[[l]]))){
    temp <- expand.grid(Comb_L[Lcomb[[l]][,k]], stringsAsFactors = FALSE)
    name <- paste(colnames(temp), collapse = "_")
    Comb_L_all[[name]] <- temp
  }
}


y_nosz1_pot <- data.frame(sed_comb = rep(names(Comb_L_all), unlist(lapply(Comb_L_all, nrow))),
                          comb_n = unlist(lapply(lapply(Comb_L_all, nrow), function(x) 1:x)),
                          sed_n = NA,
                          y_S = NA,
                          y_effN = NA)


for(i in names(Comb_L_all)){
  for(k in seq_len(nrow(Comb_L_all[[i]]))){

  # Extract OTU data from metacommunity i
    
  temp_semp <- meta[meta$MC.ID %in% Comb_L_all[[i]][k,],]$sample
  
  OTU_temp <-  OTU_nosz1[ which( rownames( OTU_nosz2) %in% temp_semp), ]
  
# Metacommunity object with weights
MC_temp <- MetaCommunity(t(OTU_temp))
  
  # Calculate phylogenetic Alpha diveristy of order q = 1, with communities weighted by there relative proportion
  y_nosz1_pot[y_nosz1_pot$sed_comb == i & y_nosz1_pot$comb_n == k ,]$y_S <- 
    GammaDiversity(MC_temp, q = 0, Correction = "ChaoWangJost")
  y_nosz1_pot[y_nosz1_pot$sed_comb == i & y_nosz1_pot$comb_n == k ,]$y_effN <- 
    GammaDiversity(MC_temp, q = 1, Correction = "Marcon")
  y_nosz1_pot[y_nosz1_pot$sed_comb == i & y_nosz1_pot$comb_n == k ,]$sed_n <- length(temp_semp)
    
  }}

y_nosz1_pot <- 
y_DIV_nosz1 %>% 
  group_by(MC.ID) %>% 
  mutate(sed_comb = paste(sediment, collapse = "_"),
         sed_n = n(), comb_n = NA) %>% 
  ungroup() %>% 
  rename( y_S = y_S_nosz1 , y_effN = y_effN_nosz1) %>% 
  select(y_S, y_effN, sed_comb, sed_n, comb_n) %>% 
  distinct(.keep_all = FALSE) %>% 
  rbind(y_nosz1_pot) %>% 
  group_by(sed_n) %>% 
  mutate(weight = n()) %>% 
  mutate(Obs = ifelse(is.na(comb_n), "observed", "potential"))

y_nosz1_pot$sed_comb <- factor(y_nosz1_pot$sed_comb, 
                               levels = y_nosz1_pot[with(y_nosz1_pot, order(sed_n, sed_comb)),]$sed_comb)

y_nosz1_pot_plot <- ggplot(y_nosz1_pot, aes(x = as.factor(sed_n), y = y_effN, fill = sed_comb, shape = Obs, alpha = Obs))+
  geom_point(size = 3, colour = "black",
             position = position_dodge(width = 0.7))+
  geom_smooth(method = "lm", se = F, aes(group = 1, weight = 1/weight), colour = "black", size = 0.3)+
  scale_fill_manual(values = c("cyan4", "forestgreen", "tan2", "brown4",
                               brewer.pal(6, "Set2"),
                               brewer.pal(4, "Dark2"),
                               brewer.pal(1, "Paired")))+
  scale_shape_manual(values = c(22,21))+
  scale_alpha_manual(values = c(1,0.6))+
  guides(fill = guide_legend(override.aes = list(shape = 21)))+
  labs(x = "Habitat richness", y = "Effective number of species", title = "nosz1")+
  theme_bw()

y_nosz1_pot_plot
  
```

## nosZ II

### Sampling depth 

The distribution of sampling depth among all pooled samples is:
```{r, cache = TRUE}
barplot(sort(rowSums(OTU_nosz2), decreasing = TRUE), col = "white")
```


### Rarefaction

```{r, cache = TRUE}
nosz2_rar <-  split(OTU_nosz2, 1:nrow(OTU_nosz2))

RareDF_nosz2 <- mclapply(nosz2_rar, function(x) {
    m <-  round(qseq(1, ifelse(sum(x) < 6000, sum(x), 6000), length.out = 30))
    rar <- iNEXT:::Dqhat.Ind(x, q = 1, m)
    out <- data.frame(m = m, qD = rar)
    return(out)})

names(RareDF_nosz2) <- row.names(OTU_nosz2)

RareDF_nosz2_df <- do.call("rbind", RareDF_nosz2)

RareDF_nosz2_df$sample <- as.numeric(gsub("(\\d+).+", "\\1", row.names(RareDF_nosz2_df)))

RareDF_nosz2_df %>% 
  left_join(meta) %>% 
  na.omit() %>% 
  ggplot(aes(x = m, y = qD, colour = sediment, group = sample))+
  facet_wrap(~Level, labeller = label_wrap_gen(multi_line=FALSE))+
  geom_line() +
  theme_bw()+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))+
  labs(y = "diversity (q = 1)", x = "sampling depth", title = "rarefaction")
```

## Alpha diversity

### Richness and effective number of species

```{r, cache = TRUE}

### Richness ###

CommunityEntropies <- apply(OTU_nosz2, 1, bcTsallis, q = 0, CheckArguments=FALSE, Correction = "ChaoWangJost")

DIV_S_nosz2 <- data.frame(sample = as.numeric(names(CommunityEntropies)), 
                       S_nosz2 = CommunityEntropies, 
                       row.names=NULL, stringsAsFactors = F)


#### Effective number of species ###

CommunityEntropies <- apply(OTU_nosz2, 1, bcTsallis, q = 1, CheckArguments=FALSE, Correction = "Marcon")
CommunityDiversity <- exp(CommunityEntropies)

DIV_effN_nosz2 <- data.frame(sample = as.numeric(names(CommunityDiversity)), 
                       effN_nosz2 = CommunityDiversity, 
                       row.names=NULL, stringsAsFactors = F)

```

### Phylogenetic diversity 

```{r, cache = TRUE}
ppTree <- Preprocess.Tree(multi2di(TREE_nosz2))


CommunityEntropies_phylo <- apply(OTU_nosz2, 1, bcPhyloEntropy, Tree = ppTree, q = 1, CheckArguments=FALSE, Correction = "Marcon")

CommunityDiversity_phylo <- unlist( lapply( CommunityEntropies_phylo, function(x) exp(x$Total)))

DIV_effN_phy_nosz2 <- data.frame(sample = as.numeric(names(CommunityDiversity_phylo)), 
                           phyloDiv_nosz2 = CommunityDiversity_phylo, row.names=NULL)

DIV_nosz2 <- 
  left_join(DIV_S_nosz2, DIV_effN_nosz2) %>% 
  left_join(DIV_effN_phy_nosz2)
  

ggscatmat(DIV_nosz2[,2:4])+
  theme_bw()
```


## Gamma diversity

In this step we pool the samples from the complete small cores.

+ For the Level 1 cores we don't need to calculate gamma diversity. We only have one pooled measurement per large core as we only have one sediment type in each level 1 core.

+ For level 2 to 4, we calculate the gamma richness, effective number of species and phylogenetic diversity as for the small cores but for the metacommunity weighted by the relative abundance of each small core.

```{r, cache = TRUE}


y_DIV_nosz2 <- data.frame(MC.ID = unique(meta[meta$Level != 1 & meta$MC.ID != "3_D", ]$MC.ID),
                    y_S_nosz2 = NA,
                    y_effN_nosz2 = NA,
                    y_phyloDiv_nosz2 = NA)


for(i in y_DIV_nosz2$MC.ID) {
  # Extract OTU data from metacommunity i
  OTU_temp <-  OTU_nosz2[ which( rownames( OTU_nosz2) %in% meta[ meta$MC.ID == i ,]$sample), ]
  
  # Metacommunity object with weights
  MC_temp <- MetaCommunity(t(OTU_temp), meta[meta$sample %in% as.numeric(rownames(OTU_temp)),]$weight)
  
  # Calculate phylogenetic Alpha diveristy of order q = 1, with communities weighted by there relative proportion
  y_DIV_nosz2[y_DIV_nosz2$MC.ID == i,]$y_S_nosz2 <- GammaDiversity(MC_temp, q = 0, Correction = "ChaoWangJost")
  y_DIV_nosz2[y_DIV_nosz2$MC.ID == i,]$y_effN_nosz2 <- GammaDiversity(MC_temp, q = 1, Correction = "Marcon")
  y_DIV_nosz2[y_DIV_nosz2$MC.ID == i,]$y_phyloDiv_nosz2 <- GammaDiversity(MC_temp, ppTree, q = 1, Correction = "Marcon")
  

}

y_DIV_nosz2 <- 
y_DIV_nosz2 %>% 
  left_join( meta, .) %>% 
  select(Level, sample, MC.ID, sediment, y_S_nosz2, y_effN_nosz2, y_phyloDiv_nosz2) %>% 
  left_join(DIV_nosz2) %>% 
  mutate(y_S_nosz2 = ifelse(Level == 1, S_nosz2, y_S_nosz2)) %>% 
  mutate(y_effN_nosz2 = ifelse(Level == 1, effN_nosz2, y_effN_nosz2)) %>% 
  mutate(y_phyloDiv_nosz2 = ifelse(Level == 1, phyloDiv_nosz2, y_phyloDiv_nosz2))

write.table(y_DIV_nosz2, "DIV_nosz2.txt")

```
## potential Gamma diversity

```{r}

meta1 <- filter(meta, Level == 1, replicate != "A")

Comb_L <- split(meta1$MC.ID, meta1$sediment)

Lcomb <- list(combn(4,2), combn(4,3), combn(4,4))

Comb_L_all <- list()

for (l in seq_along(Lcomb)){
  for (k in seq_len(ncol(Lcomb[[l]]))){
    temp <- expand.grid(Comb_L[Lcomb[[l]][,k]], stringsAsFactors = FALSE)
    name <- paste(colnames(temp), collapse = "_")
    Comb_L_all[[name]] <- temp
  }
}


y_nosz2_pot <- data.frame(sed_comb = rep(names(Comb_L_all), unlist(lapply(Comb_L_all, nrow))),
                          comb_n = unlist(lapply(lapply(Comb_L_all, nrow), function(x) 1:x)),
                          sed_n = NA,
                          y_S = NA,
                          y_effN = NA)


for(i in names(Comb_L_all)){
  for(k in seq_len(nrow(Comb_L_all[[i]]))){

  # Extract OTU data from metacommunity i
    
  temp_semp <- meta[meta$MC.ID %in% Comb_L_all[[i]][k,],]$sample
  
  OTU_temp <-  OTU_nosz2[ which( rownames( OTU_nosz2) %in% temp_semp), ]
  
# Metacommunity object with weights
MC_temp <- MetaCommunity(t(OTU_temp))
  
  # Calculate phylogenetic Alpha diveristy of order q = 1, with communities weighted by there relative proportion
  y_nosz2_pot[y_nosz2_pot$sed_comb == i & y_nosz2_pot$comb_n == k ,]$y_S <- 
    GammaDiversity(MC_temp, q = 0, Correction = "ChaoWangJost")
  y_nosz2_pot[y_nosz2_pot$sed_comb == i & y_nosz2_pot$comb_n == k ,]$y_effN <- 
    GammaDiversity(MC_temp, q = 1, Correction = "Marcon")
  y_nosz2_pot[y_nosz2_pot$sed_comb == i & y_nosz2_pot$comb_n == k ,]$sed_n <- length(temp_semp)
    
}}

#filter(Level == 1, ! grepl("A", MC.ID)) %>% 

y_nosz2_pot <- 
y_DIV_nosz2 %>% 
  group_by(MC.ID) %>% 
  mutate(sed_comb = paste(sediment, collapse = "_"),
         sed_n = n(), comb_n = NA) %>% 
  ungroup() %>% 
  rename( y_S = y_S_nosz2 , y_effN = y_effN_nosz2) %>% 
  select(y_S, y_effN, sed_comb, sed_n, comb_n) %>% 
  distinct(.keep_all = FALSE) %>% 
  rbind(y_nosz2_pot) %>% 
  group_by(sed_n) %>% 
  mutate(weight = n()) %>% 
  mutate(Obs = ifelse(is.na(comb_n), "observed", "potential"))

y_nosz2_pot$sed_comb <- factor(y_nosz2_pot$sed_comb, 
                               levels = y_nosz2_pot[with(y_nosz2_pot, order(sed_n, sed_comb)),]$sed_comb)

y_nosz2_pot_plot <- ggplot(y_nosz2_pot,aes(x = as.factor(sed_n), y = y_effN, fill = sed_comb, shape = Obs, alpha = Obs))+
  geom_point(size = 3, colour = "black",
             position = position_dodge(width = 0.7))+
  geom_smooth(method = "lm", se = F, aes(group = 1, weight = 1/weight), colour = "black", size = 0.3)+
  scale_fill_manual(values = c("cyan4", "forestgreen", "tan2", "brown4",
                               brewer.pal(6, "Set2"),
                               brewer.pal(4, "Dark2"),
                               brewer.pal(1, "Paired")))+
  scale_shape_manual(values = c(22,21))+
  scale_alpha_manual(values = c(1,0.6))+
  guides(fill = guide_legend(override.aes = list(shape = 21)))+
  labs(x = "Habitat richness", y = "Effective number of species", title = "nosZ")+
  theme_bw()

y_nosz2_pot_plot
```

```{r, fig.height=10, fig.width=12}
y_all_pot <- y_nosz2_pot %>% 
  mutate(gene = "nosz2") %>% 
  rbind(mutate(y_nosz1_pot, gene = "nosz1")) %>%
  rbind(mutate(y_nifh_pot, gene = "nifh")) %>%
  rbind(mutate(y_bac_pot, gene = "16S")) 

y_all_pot$sed_comb <- factor(y_all_pot$sed_comb,
                               levels = y_all_pot[with(y_all_pot, order(gene, sed_n, sed_comb)),]$sed_comb)


pot_g_plot <- ggplot(y_all_pot,aes(x = as.factor(sed_n), y = y_effN, fill = sed_comb, shape = Obs, alpha = Obs))+
  facet_wrap(~gene, scales = "free_y", nrow = 2)+
  geom_point(size = 3, colour = "black",
             position = position_dodge(width = 0.7))+
  geom_smooth(method = "lm", se = F, aes(group = 1, weight = 1/weight), colour = "black", size = 0.3)+
  scale_fill_manual(values = c("cyan4", "forestgreen", "tan2", "brown4",
                               brewer.pal(6, "Set2"),
                               brewer.pal(4, "Dark2"),
                               brewer.pal(1, "Paired")))+
  scale_shape_manual(values = c(22,21))+
  scale_alpha_manual(values = c(1,0.6))+
  guides(fill = guide_legend(override.aes = list(shape = 21), title = ""))+
  labs(x = "Habitat richness", y = "Effective number of species", title = "Gamma diversity")+
  theme_bw(base_size = 15)+
  theme(legend.position = "bottom")

pot_g_plot 

ggsave("potential_gamma_diversity.pdf", height = 12, width = 14)
ggsave("potential_gamma_diversity.png", height = 12, width = 14)
```


## Joining and exporting the data

```{r}
DIV <- left_join(y_DIV_nifh, y_DIV_nosz1) %>% left_join(., y_DIV_nosz2) %>% left_join(., y_DIV_bac)

write.table(DIV, "Diversity_estimations.txt")
```

# Alpha diversity plots

```{r}
DIV_plot <- DIV %>% 
  select(sample, matches("^effn|^phylo|^S_")) %>%
  na.omit() %>% 
  gather(variable, value, -sample) %>% 
  mutate(gene = gsub("\\w+_(\\w+)", "\\1", variable)) %>%
  mutate(metric = gsub("(\\w+)_\\w+", "\\1", variable)) %>% 
  select(-variable) %>% 
  spread(metric, value) %>% 
  left_join(meta)
  
```

```{r}
ggplot(DIV_plot, aes(x = sediment, y = S, colour = sediment, shape = as.factor(Level)))+
  geom_point(position = position_dodge(width = 0.5))+
  facet_wrap(~gene, scale = "free_y")+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme_bw()+
  theme(axis.text.x=element_text(angle = -90, hjust = 0, vjust = 0.5))+
  labs(title = "species richness")
```

```{r}
ggplot(DIV_plot, aes(x = sediment, y = effN, colour = sediment, shape = as.factor(Level)))+
  geom_point(position = position_dodge(width = 0.5))+
  facet_wrap(~gene, scale = "free_y")+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme_bw()+
  theme(axis.text.x=element_text(angle = -90, hjust = 0, vjust = 0.5))+
  labs(title = "effective number of species")
```


```{r}
ggplot(DIV_plot, aes(x = sediment, y = phyloDiv, colour = sediment, shape = as.factor(Level)))+
  geom_point(position = position_dodge(width = 0.5))+
  facet_wrap(~gene, scale = "free_y")+
  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme_bw()+
  theme(axis.text.x=element_text(angle = -90, hjust = 0, vjust = 0.5))+
  labs(title = "phylogenetic diversity")
```

# Gamma diversity plots

```{r}

DIV_plot_y <- DIV %>% 
  select(sample, matches("^y")) %>%
  na.omit() %>% 
  gather(variable, value, -sample) %>% 
  mutate(gene = gsub("y_\\w+_(\\w+)", "\\1", variable)) %>%
  mutate(metric = gsub("(y_\\w+)_\\w+", "\\1", variable)) %>% 
  select(-variable) %>% 
  spread(metric, value) %>% 
  left_join(meta) %>% 
  mutate(sediment = as.numeric(as.factor(sediment))) %>% 
  group_by(gene, replicate, Level, season) %>% 
  summarize_all("mean") 

```

```{r}
ggplot(DIV_plot, aes(x = sediment, y = S))+
  geom_point(alpha = 0.8, size = 3, shape = 21, aes(colour = replicate), fill = NA)+
  geom_line(data = filter(DIV_plot, Level > 1), aes(x = sediment, y = S, group = replicate, colour = replicate), alpha = 0.4)+
  geom_point(data = filter(DIV_plot_y, Level > 1), aes( x = sediment, y = y_S, fill = replicate), shape = 23, size = 3, alpha = 0.7)+
  facet_grid(gene~Level, scale = "free")+
 # scale_fill_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
#  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme_bw()+
  theme(axis.text.x=element_text(angle = -90, hjust = 0, vjust = 0.5))+
  labs(title = "species richness (gamma)")
```

```{r}
ggplot(DIV_plot, aes(x = sediment, y = effN))+
  geom_point(alpha = 0.8, size = 3, shape = 21, aes(colour = replicate), fill = NA)+
  geom_line(data = filter(DIV_plot, Level > 1), aes(x = sediment, y = effN, group = replicate, colour = replicate), alpha = 0.4)+
  geom_point(data = filter(DIV_plot_y, Level > 1), aes( x = sediment, y = y_effN, fill = replicate), shape = 23, size = 3, alpha = 0.7)+
  facet_grid(gene~Level, scale = "free")+
 # scale_fill_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
#  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme_bw()+
  theme(axis.text.x=element_text(angle = -90, hjust = 0, vjust = 0.5))+
  labs(title = "effective number of species (gamma)")
```


```{r}
ggplot(DIV_plot, aes(x = sediment, y = phyloDiv))+
  geom_point(alpha = 0.8, size = 3, shape = 21, aes(colour = replicate), fill = NA)+
  geom_line(data = filter(DIV_plot, Level > 1), aes(x = sediment, y = phyloDiv, group = replicate, colour = replicate), alpha = 0.4)+
  geom_point(data = filter(DIV_plot_y, Level > 1), aes( x = sediment, y = y_phyloDiv, fill = replicate), shape = 23, size = 3, alpha = 0.7)+
  facet_grid(gene~Level, scale = "free")+
 # scale_fill_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
#  scale_colour_manual(values = c("cyan4", "forestgreen", "tan2", "brown4"))+
  theme_bw()+
  theme(axis.text.x=element_text(angle = -90, hjust = 0, vjust = 0.5))+
  labs(title = "phylogenetic diversity")
```

save workspace
```{r}
save.image("Div_est.RData")
#load("Div_est.RData")
```

